tool_path_planner/src/utilities.cpp

```diff
@@ -484,6 +503,165 @@ ToolPaths addExtraPaths(const ToolPaths& tool_paths, double offset_distance)
+    -1 * axis_1 + -1 * axis_2,
+    -1 * axis_1 +      axis_2,
+         axis_1 +      axis_2,
+         axis_1 + -1 * axis_2
+  };
+  // Get indices to cut tool path at
+  std::set<int> cut_indices;
+
+  Eigen::Vector3f path_center(0.0, 0.0, 0.0);
+  for (std::size_t index = 0; index < tool_path_segment.size(); ++index)
+  {
+    Eigen::Isometry3d p = tool_path_segment[index];
+    path_center[0] += p.translation().x();
+    path_center[1] += p.translation().y();
+    path_center[2] += p.translation().z();
+  }
+
+  path_center = path_center / tool_path_segment.size();
+
+  for (Eigen::Vector3f vector : vectors)
+  {
+    float max_dot = std::numeric_limits<float>::min();
+    int max_index = -1;
+    for (std::size_t index = 0; index < tool_path_segment.size(); ++index)
+    {
+      Eigen::Isometry3d p = tool_path_segment[index];
+      Eigen::Vector3f v(float(p.translation().x() - path_center[0]), float(p.translation().y() - path_center[1]), float(p.translation().z() - path_center[2]));
+      float dot = vector.dot(v);
+      if (dot > max_dot)
+      {
+        max_dot = dot;
+        max_index = int(index);
+      }
+    }
+    cut_indices.insert(max_index);
+  }
+  // Cut segment into smaller segments (at cut indices)
+  for (std::set<int>::iterator i = cut_indices.begin(); i != cut_indices.end(); ++i)
+  {
+    ToolPathSegment new_tool_path_segment;
+    int current_index = *i;
+    int next_index;
+    if (std::next(i, 1) == cut_indices.end())
+    {
+      next_index = *cut_indices.begin();
+    }
+    else
+    {
+      next_index = *std::next(i, 1);
+    }
+    while (current_index != next_index)
+    {
+      new_tool_path_segment.push_back(tool_path_segment[std::size_t(current_index)]);
+      current_index = (current_index + 1) % int(tool_path_segment.size());
+    }
+    new_tool_path.push_back(new_tool_path_segment);
+  }
+  return new_tool_path;
+}
+
+ToolPaths splitByAxes(const ToolPaths& tool_paths)
+{
+  if (tool_paths.size() == 0)
+  {
+    ROS_ERROR_STREAM("Tool paths is empty.");
+    // TODO: throw an exception
+  }
+
+  using Point = pcl::PointXYZ;
+  ToolPaths new_tool_paths;
+
+  for (std::size_t tool_path_index = 0; tool_path_index < tool_paths.size(); ++tool_path_index)
+  {
+    ToolPath tool_path = tool_paths[tool_path_index];
+    // Sanity check - tool path must have exactly one segment
+    if (tool_path.size() != 1)
+    {
+      ROS_ERROR_STREAM("Tool path " << tool_path_index << " contains " << tool_path.size() <<
+                       " segments. Expected exactly 1.");
+      // TODO: throw an exception
+    }
+    ToolPath new_tool_path = splitByAxes(tool_path[0]);
+    new_tool_paths.push_back(new_tool_path);
+  }
+  return new_tool_paths;
+}
+
+ToolPaths splitByAxes(const ToolPaths& tool_paths, const Eigen::Vector3f& axis_1, const Eigen::Vector3f& axis_2)
+{
+  if (tool_paths.size() == 0)
+  {
+    ROS_ERROR_STREAM("Tool paths is empty.");
+    // TODO: throw an exception
```

Comment:
Still don't like these "TODO: Exception".  Don't like the unresolved TODOs, don't like that the proposed solution is an exception.  Just return something empty.